File: src/sniff/rules.py
import re
from collections.abc import Iterable
from pathlib import Path
from typing import Any

import yaml

from sniff.models import Rule, Sniffer, SniffError

SEVERITIES = {"suggestion", "warning", "error"}
SNIFFER_KINDS = {"llm", "vale", "ruff"}
RULE_ID = re.compile(r"^[a-z][a-z0-9]*(?:-[a-z0-9]+)+$")
RULE_CODE = re.compile(r"^[A-Z]{3}[0-9]{3}$")


def _read_rule(path: Path) -> Rule:
    text = path.read_text(encoding="utf-8")
    if not text.startswith("---\n"):
        raise SniffError(f"{path}: rule must start with YAML frontmatter")
    try:
        raw_frontmatter, body = text[4:].split("\n---\n", 1)
    except ValueError as exc:
        raise SniffError(f"{path}: unterminated YAML frontmatter") from exc
    try:
        data = yaml.safe_load(raw_frontmatter)
    except yaml.YAMLError as exc:
        raise SniffError(f"{path}: invalid YAML: {exc}") from exc
    if not isinstance(data, dict):
        raise SniffError(f"{path}: frontmatter must be a mapping")

    rule_id = data.get("id")
    if not isinstance(rule_id, str) or not RULE_ID.fullmatch(rule_id):
        raise SniffError(f"{path}: invalid rule id {rule_id!r}")
    if rule_id != path.stem:
        raise SniffError(f"{path}: id must equal filename stem")

    code = data.get("code")
    if not isinstance(code, str) or not RULE_CODE.fullmatch(code):
        raise SniffError(f"{path}: invalid rule code {code!r}; expected ABC123")

    applies_to = data.get("applies_to")
    if (
        not isinstance(applies_to, list)
        or not applies_to
        or not all(isinstance(item, str) and item for item in applies_to)
    ):
        raise SniffError(f"{path}: applies_to must be a non-empty string list")

    severity = data.get("severity")
    if severity not in SEVERITIES:
        raise SniffError(f"{path}: severity must be suggestion, warning, or error")

    raw_sniffers = data.get("sniffers")
    if not isinstance(raw_sniffers, list) or not raw_sniffers:
        raise SniffError(f"{path}: sniffers must be a non-empty list")
    sniffers: list[Sniffer] = []
    for raw in raw_sniffers:
        if not isinstance(raw, dict) or raw.get("kind") not in SNIFFER_KINDS:
            raise SniffError(f"{path}: unsupported sniffer {raw!r}")
        kind = raw["kind"]
        options = {key: value for key, value in raw.items() if key != "kind"}
        if kind == "vale" and not isinstance(options.get("pattern"), str):
            raise SniffError(f"{path}: vale sniffer requires pattern")
        if kind == "ruff" and not isinstance(options.get("rule"), str):
            raise SniffError(f"{path}: ruff sniffer requires rule")
        sniffers.append(Sniffer(kind=kind, options=options))

    llm_exempt = bool(data.get("llm_exempt", False))
    if not llm_exempt and not any(item.kind == "llm" for item in sniffers):
        raise SniffError(f"{path}: add an llm sniffer or set llm_exempt: true")

    required_strings = ("name", "family", "message")
    for key in required_strings:
        if not isinstance(data.get(key), str) or not data[key].strip():
            raise SniffError(f"{path}: {key} must be a non-empty string")

    return Rule(
        id=rule_id,
        code=code,
        name=data["name"],
        family=data["family"],
        applies_to=tuple(applies_to),
        severity=severity,
        message=data["message"],
        body=body.strip(),
        path=path,
        sniffers=tuple(sniffers),
        fix=str(data.get("fix", "manual")),
        evidence=str(data.get("evidence", "")),
        llm_exempt=llm_exempt,
    )


def load_rules(directories: Iterable[Path]) -> dict[str, Rule]:
    loaded: dict[str, Rule] = {}
    codes: dict[str, Rule] = {}
    for directory in directories:
        if not directory.is_dir():
            continue
        for path in sorted(directory.glob("*.md")):
            rule = _read_rule(path)
            if rule.id in loaded:
                raise SniffError(
                    f"duplicate rule id {rule.id!r}: {loaded[rule.id].path} and {path}"
                )
            if rule.code in codes:
                raise SniffError(
                    f"duplicate rule code {rule.code!r}: {codes[rule.code].path} and {path}"
                )
            loaded[rule.id] = rule
            codes[rule.code] = rule
    return loaded


def selected_rules(
    rules: dict[str, Rule], config: dict[str, Any], profile_name: str
) -> list[Rule]:
    profiles = config.get("profiles", {})
    profile = profiles.get(profile_name)
    if not isinstance(profile, dict):
        choices = ", ".join(sorted(profiles))
        raise SniffError(f"unknown profile {profile_name!r}; choose one of: {choices}")
    tags = set(profile.get("tags", []))
    include_ids = set(profile.get("include_rules", []))
    exclude_ids = set(profile.get("exclude_rules", []))
    overrides = config.get("rules", {})
    referenced = include_ids | exclude_ids | set(overrides)
    unknown = referenced - set(rules)
    if unknown:
        raise SniffError(
            f"configuration references unknown rule(s): {', '.join(sorted(unknown))}"
        )

    chosen: list[Rule] = []
    for rule in rules.values():
        override = overrides.get(rule.id, {}) if isinstance(overrides, dict) else {}
        if isinstance(override, dict) and override.get("enabled") is False:
            continue
        applies = "*" in tags or bool(tags.intersection(rule.applies_to))
        if rule.id in include_ids:
            applies = True
        if rule.id in exclude_ids:
            applies = False
        if applies:
            chosen.append(rule)
    return sorted(chosen, key=lambda item: item.id)


def effective_severity(rule: Rule, config: dict[str, Any], profile_name: str) -> str:
    severity = rule.severity
    profile = config.get("profiles", {}).get(profile_name, {})
    profile_severity = profile.get("severity", {}) if isinstance(profile, dict) else {}
    if isinstance(profile_severity, dict):
        severity = profile_severity.get(rule.id, severity)
    override = config.get("rules", {}).get(rule.id, {})
    if isinstance(override, dict):
        severity = override.get("severity", severity)
    if severity not in SEVERITIES:
        raise SniffError(f"{rule.id}: invalid effective severity {severity!r}")
    return severity

File: rust/rules.rs
use crate::{
    error::{Result, message},
    models::{Context, RawRule, Rule, Severity},
};
use include_dir::{Dir, include_dir};
use regex::Regex;
use std::{
    collections::{BTreeMap, BTreeSet},
    fs,
};

static BUNDLED: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/rules");

fn parse(source: &str, text: &str, stem: &str) -> Result<Rule> {
    let rest = text
        .strip_prefix("---\n")
        .ok_or_else(|| message(format!("{source}: rule must start with YAML frontmatter")))?;
    let (front, body) = rest
        .split_once("\n---\n")
        .ok_or_else(|| message(format!("{source}: unterminated YAML frontmatter")))?;
    let raw: RawRule = serde_yaml_ng::from_str(front)
        .map_err(|e| message(format!("{source}: invalid YAML: {e}")))?;
    if !Regex::new(r"^[a-z][a-z0-9]*(?:-[a-z0-9]+)+$")
        .unwrap()
        .is_match(&raw.id)
    {
        return Err(message(format!("{source}: invalid rule id {:?}", raw.id)));
    }
    if raw.id != stem {
        return Err(message(format!("{source}: id must equal filename stem")));
    }
    if !Regex::new(r"^[A-Z]{3}[0-9]{3}$")
        .unwrap()
        .is_match(&raw.code)
    {
        return Err(message(format!(
            "{source}: invalid rule code {:?}; expected ABC123",
            raw.code
        )));
    }
    if raw.applies_to.is_empty() {
        return Err(message(format!(
            "{source}: applies_to must be a non-empty string list"
        )));
    }
    if raw.sniffers.is_empty() {
        return Err(message(format!(
            "{source}: sniffers must be a non-empty list"
        )));
    }
    for sniffer in &raw.sniffers {
        if !["llm", "vale", "ruff"].contains(&sniffer.kind.as_str()) {
            return Err(message(format!(
                "{source}: unsupported sniffer {:?}",
                sniffer.kind
            )));
        }
        let required = match sniffer.kind.as_str() {
            "vale" => Some("pattern"),
            "ruff" => Some("rule"),
            _ => None,
        };
        if required.is_some_and(|k| {
            !matches!(
                sniffer.options.get(k),
                Some(serde_yaml_ng::Value::String(_))
            )
        }) {
            return Err(message(format!(
                "{source}: {} sniffer requires {}",
                sniffer.kind,
                required.unwrap()
            )));
        }
    }
    if !raw.llm_exempt && !raw.sniffers.iter().any(|s| s.kind == "llm") {
        return Err(message(format!(
            "{source}: add an llm sniffer or set llm_exempt: true"
        )));
    }
    Ok(Rule {
        id: raw.id,
        code: raw.code,
        name: raw.name,
        family: raw.family,
        applies_to: raw.applies_to,
        severity: raw.severity,
        message: raw.message,
        body: body.trim().into(),
        source: source.into(),
        sniffers: raw.sniffers,
        fix: raw.fix,
        evidence: raw.evidence,
        llm_exempt: raw.llm_exempt,
    })
}

pub fn load(context: &Context) -> Result<BTreeMap<String, Rule>> {
    let mut rules = BTreeMap::new();
    let mut codes = BTreeMap::<String, String>::new();
    for file in BUNDLED
        .files()
        .filter(|f| f.path().extension().is_some_and(|x| x == "md"))
    {
        let source = file.path().display().to_string();
        let text = file.contents_utf8().unwrap();
        let rule = parse(
            &source,
            text,
            file.path().file_stem().unwrap().to_str().unwrap(),
        )?;
        insert(&mut rules, &mut codes, rule)?;
    }
    let project_rules = context.project_root.join(".sniff/rules");
    for directory in context
        .rule_dirs
        .iter()
        .chain(std::iter::once(&project_rules))
    {
        if !directory.is_dir() {
            continue;
        }
        let mut paths: Vec<_> = fs::read_dir(directory)
            .map_err(|e| message(e.to_string()))?
            .filter_map(|e| e.ok().map(|x| x.path()))
            .filter(|p| p.extension().is_some_and(|x| x == "md"))
            .collect();
        paths.sort();
        for path in paths {
            let text = fs::read_to_string(&path)
                .map_err(|e| message(format!("{}: {e}", path.display())))?;
            let rule = parse(
                &path.display().to_string(),
                &text,
                path.file_stem().unwrap().to_str().unwrap(),
            )?;
            insert(&mut rules, &mut codes, rule)?;
        }
    }
    Ok(rules)
}
fn insert(
    rules: &mut BTreeMap<String, Rule>,
    codes: &mut BTreeMap<String, String>,
    rule: Rule,
) -> Result<()> {
    if let Some(old) = rules.get(&rule.id) {
        return Err(message(format!(
            "duplicate rule id {:?}: {} and {}",
            rule.id, old.source, rule.source
        )));
    }
    if let Some(old) = codes.get(&rule.code) {
        return Err(message(format!(
            "duplicate rule code {:?}: {} and {}",
            rule.code, old, rule.source
        )));
    }
    codes.insert(rule.code.clone(), rule.source.clone());
    rules.insert(rule.id.clone(), rule);
    Ok(())
}
fn table<'a>(v: &'a toml::Value, key: &str) -> Option<&'a toml::map::Map<String, toml::Value>> {
    v.get(key)?.as_table()
}
fn strings(v: Option<&toml::Value>) -> BTreeSet<String> {
    v.and_then(|x| x.as_array())
        .into_iter()
        .flatten()
        .filter_map(|x| x.as_str().map(str::to_owned))
        .collect()
}
pub fn selected<'a>(
    all: &'a BTreeMap<String, Rule>,
    data: &toml::Value,
    profile: &str,
) -> Result<Vec<&'a Rule>> {
    let profiles = table(data, "profiles").unwrap();
    let p = profiles
        .get(profile)
        .and_then(|v| v.as_table())
        .ok_or_else(|| {
            message(format!(
                "unknown profile {profile:?}; choose one of: {}",
                profiles.keys().cloned().collect::<Vec<_>>().join(", ")
            ))
        })?;
    let tags = strings(p.get("tags"));
    let include = strings(p.get("include_rules"));
    let exclude = strings(p.get("exclude_rules"));
    let overrides = table(data, "rules");
    let referenced: BTreeSet<_> = include
        .iter()
        .chain(&exclude)
        .chain(overrides.into_iter().flat_map(|x| x.keys()))
        .cloned()
        .collect();
    let unknown: Vec<_> = referenced
        .difference(&all.keys().cloned().collect())
        .cloned()
        .collect();
    if !unknown.is_empty() {
        return Err(message(format!(
            "configuration references unknown rule(s): {}",
            unknown.join(", ")
        )));
    }
    Ok(all
        .values()
        .filter(|r| {
            let o = overrides
                .and_then(|x| x.get(&r.id))
                .and_then(|x| x.as_table());
            if o.and_then(|x| x.get("enabled")).and_then(|x| x.as_bool()) == Some(false) {
                return false;
            }
            let mut yes = tags.contains("*") || r.applies_to.iter().any(|x| tags.contains(x));
            if include.contains(&r.id) {
                yes = true
            };
            if exclude.contains(&r.id) {
                yes = false
            };
            yes
        })
        .collect())
}
pub fn severity(rule: &Rule, data: &toml::Value, profile: &str) -> Result<Severity> {
    let mut s = rule.severity;
    if let Some(x) = data
        .get("profiles")
        .and_then(|x| x.get(profile))
        .and_then(|x| x.get("severity"))
        .and_then(|x| x.get(&rule.id))
        .and_then(|x| x.as_str())
    {
        s = Severity::parse(x)
            .ok_or_else(|| message(format!("{}: invalid effective severity {x:?}", rule.id)))?
    };
    if let Some(x) = data
        .get("rules")
        .and_then(|x| x.get(&rule.id))
        .and_then(|x| x.get("severity"))
        .and_then(|x| x.as_str())
    {
        s = Severity::parse(x)
            .ok_or_else(|| message(format!("{}: invalid effective severity {x:?}", rule.id)))?
    };
    Ok(s)
}
